File: backend/src/main/java/iuh/fit/backend/model/ProductVariant.java
package iuh.fit.backend.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.util.List;

@Entity
@Table(name = "product_variants")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class ProductVariant {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Version
    private Long version;  // Optimistic locking for additional safety

    @Column(name = "product_id")
    private Integer productId;

    private String name;
    private String sku;
    private BigDecimal price;

    @Column(name = "sale_price")
    private BigDecimal salePrice;

    @Column(name = "stock_quantity")
    private Integer stockQuantity;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", insertable = false, updatable = false)
    @JsonIgnore
    private Product product;

    @OneToMany(mappedBy = "productVariant", fetch = FetchType.LAZY)
    @JsonIgnore
    private List<VariantAttribute> variantAttributes;

    @OneToOne(mappedBy = "productVariant", fetch = FetchType.LAZY)
    @JsonIgnore
    private CartItem cartItem;

    @OneToOne(mappedBy = "productVariant", fetch = FetchType.LAZY)
    @JsonIgnore
    private OrderItem orderItem;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "favorite_id", insertable = false, updatable = false)
    private FavoriteList favoriteList;
}

File: frontend/src/types/ProductVariant.ts
import type { Product } from './Product';
import type { VariantAttribute } from './VariantAttribute';
import type { CartItem } from './CartItem';
import type { OrderItem } from './OrderItem';
import type { FavoriteList } from './FavoriteList';

export interface ProductVariant {
  id: number;
  productId: number;
  name: string;
  sku: string;
  price: number;
  salePrice: number;
  stockQuantity: number;
  product?: Product;
  variantAttributes?: VariantAttribute[];
  cartItem?: CartItem;
  orderItem?: OrderItem;
  favoriteList?: FavoriteList;
}
